crates/types/src/storage/decode.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::mem;
+
+use bytes::Buf;
+use serde::de::{DeserializeOwned, Error as DeserializationError};
+use tracing::error;
+
+use super::{StorageCodecKind, StorageDecodeError};
+
+/// Decode a [`DeserializeOwned`] type from a buffer using serde if it is supported by the codec.
+pub fn decode_serde<T: DeserializeOwned, B: Buf>(
+    buf: &mut B,
+    codec: StorageCodecKind,
+) -> Result<T, StorageDecodeError> {
+    match codec {
+        StorageCodecKind::FlexbuffersSerde => decode_serde_from_flexbuffers(buf, true)
+            .map_err(|err| StorageDecodeError::DecodeValue(err.into())),
+        StorageCodecKind::FlexbuffersNoPrefix => decode_serde_from_flexbuffers(buf, false)
+            .map_err(|err| StorageDecodeError::DecodeValue(err.into())),
+        StorageCodecKind::BincodeSerde => decode_serde_from_bincode(buf)
+            .map_err(|err| StorageDecodeError::DecodeValue(err.into())),
+        StorageCodecKind::Json => {
+            decode_serde_from_json(buf).map_err(|err| StorageDecodeError::DecodeValue(err.into()))
+        }
+        codec => Err(StorageDecodeError::UnsupportedCodecKind(codec)),
+    }
+}
+
+/// Utility method to decode a [`DeserializeOwned`] type from flexbuffers using serde.
+fn decode_serde_from_flexbuffers<T: DeserializeOwned, B: Buf>(
+    buf: &mut B,
+    length_prefix: bool,
+) -> Result<T, flexbuffers::DeserializationError> {
+    if !length_prefix {
+        // without length-prefix, we rely on the buffer being contiguous.
+        let deserializer = flexbuffers::Reader::get_root(buf.chunk())?;
+        let result = serde_path_to_error::deserialize(deserializer).map_err(|err| {
+            error!(%err, "Flexbuffers error at field {}", err.path());
+            err.into_inner()
+        })?;
+        return Ok(result);
+    }
+
+    if buf.remaining() < mem::size_of::<u32>() {
+        return Err(flexbuffers::DeserializationError::custom(format!(
+            "insufficient data: expecting {} bytes for length",
+            mem::size_of::<u32>()
+        )));
+    }
+    let length = usize::try_from(buf.get_u32_le()).expect("u32 to fit into usize");
+
+    if buf.remaining() < length {
+        return Err(flexbuffers::DeserializationError::custom(format!(
+            "insufficient data: expecting {length} bytes for flexbuffers"
+        )));
+    }
+
+    if buf.chunk().len() >= length {
+        let deserializer = flexbuffers::Reader::get_root(buf.chunk())?;
```

Comment:
This might have been broken for quite some time.

```suggestion
        let deserializer = flexbuffers::Reader::get_root(&buf.chunk()[..length])?;
```

Reply:
The following test is indeed failing w/o this change:

```
#[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
    struct Foobar {
        value: String,
    }

    flexbuffers_storage_encode_decode!(Foobar);

    #[test]
    fn test_multiple_flexbuffers() {
        let mut buffer = BytesMut::new();

        let a = Foobar {
            value: "a".to_owned(),
        };
        let b = Foobar {
            value: "b".to_owned(),
        };

        StorageCodec::encode(&a, &mut buffer).unwrap();
        StorageCodec::encode(&b, &mut buffer).unwrap();

        let mut buffer = buffer.freeze();

        let a_r = StorageCodec::decode(&mut buffer).unwrap();
        let b_r = StorageCodec::decode(&mut buffer).unwrap();

        assert_eq!(a, a_r);
        assert_eq!(b, b_r);
    }
```

Reply:
I think this is not the case right now. That's why we haven't seen this problem yet. I guess it boils down to the question whether this is a hard assumption we want to make. With the current API one might get surprised by it.

Reply:
I'll remove the non-length prefixed version from flexbuffers as I don't think I'll actually use it. 

Let's split the change into two, simple one that adds support for decoding non-length prefixed bincode+json to get ready for v1.3. Then let's discuss separately any API changes that would make it more obvious that we expect the input to have a single value.